src/project/git/StableDirectoryManager.ts

```diff
@@ -0,0 +1,135 @@
+import { registerShutdownHook } from "../../internal/util/shutdown";
+
+import { logger } from "../../internal/util/logger";
+
+import * as fs from "fs-extra";
+import * as os from "os";
+import { CloneDirectoryInfo, CloneOptions, DirectoryManager } from "./DirectoryManager";
+
+export interface StableDirectoryManagerOpts {
+
+    /**
+     * Maximum number of directories to use
+     * @type {number}
+     */
+    maxDirectories?: number;
+
+    /**
+     * Attempt to reuse directories?
+     */
+    reuseDirectories: boolean;
+
+    cleanOnExit?: boolean;
+
+    /**
+     * If provided, number of directories
+     */
+    baseDir: string;
+}
+
+const DefaultOpts = {
+
+    maxDirectories: 100,
+    cleanOnExit: true,
+};
+
+/**
+ * Build a stable working directory structure.
+ * Options determine whether it's cleared on exit.
+ */
+export class StableDirectoryManager implements DirectoryManager {
+
+    public opts: StableDirectoryManagerOpts;
+
+    private directoriesUsed = 0;
+
+    constructor(pOpts: StableDirectoryManagerOpts) {
+        this.opts = {
+            ...DefaultOpts,
+            ...pOpts,
+        };
+
+        logger.info("Setting up StableDirectoryManager with options %s", JSON.stringify(this.opts));
```

Comment:
debug? also have you seen `logger.info("Setting up StableDirectoryManager with options %j"`? see https://nodejs.org/api/util.html#util_util_format_format_args. I only found about this recently.

Reply:
Thanks!